### PR TITLE
Bug/Composer page is not working

### PR DESCRIPTION
**Pull request by eerison**
## Subject

I am targeting this branch, because {reason}.

Closes #{put_issue_number_here}.

## Changelog

```markdown
### Added
- Added some `Class::newMethod()` to do great stuff.

### Changed

### Deprecated

### Removed

### Fixed

### Security
```

### CODE REVIEW

File: src/Controller/PageAdminController.php [hunk 1]
```diff
@@ -294,6 +294,7 @@ public function composeContainerShowAction(Request $request): Response
             }
             $container = $template->getContainer($blockCode);
 
+            //TODO is the problem, in 3.x block services have items, but in 4.x it's empty.
```
Comment (eerison) [inline src/Controller/PageAdminController.php]: Hey @VincentLanglet in this action I basically just need the `getObject`, from [AdminInterface](https://github.com/sonata-project/SonataAdminBundle/blob/4.x/src/Admin/AdminInterface.php#L221), could we create an interface with this? I guess inject the whole AdminInterface to use only one method is a lot, WDYT?

File: src/Admin/PageAdmin.php [hunk 2]
```diff
@@ -62,6 +62,7 @@ public function __construct(
     protected function configureRoutes(RouteCollectionInterface $collection): void
     {
         $collection->add('compose', $this->getRouterIdParameter().'/compose');
+        //TODO how is it linked with src/Controller/PageAdminController::composeContainerShowAction, where the route is defined?
```
Comment (eerison) [inline src/Admin/PageAdmin.php]: do you have any idea where this router is defined? `compose_container_show`

Other question: WDYT to  extract this [composeContainerShowAction](https://github.com/sonata-project/SonataPageBundle/blob/4.x/src/Controller/PageAdminController.php#L257) for other controller like `PageAdminComposeContainerShowController` or just `PageAdminComposeController` with `composeContainerShowAction` and `composerAction`?

**Comment by jordisala1991** [comment]
What is not working? maybe if you add a failing test first it would be easier to see. (I tried the compose page quite a lot and fixed all the errors there)

**Comment by eerison** [comment]
> What is not working? maybe if you add a failing test first it would be easier to see. (I tried the compose page quite a lot and fixed all the errors there)

the problem posted here: https://github.com/sonata-project/SonataPageBundle/issues/1495#issuecomment-1223876198

and yes, I'm trying to add a failing test first, to find out what exactly is broken.

File: src/Admin/PageAdmin.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Admin/PageAdmin.php]: I would say It's just linked with the url you gave `compose/container/{id}`

File: src/Controller/PageAdminController.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Controller/PageAdminController.php]: I don't see a reason to create another interface. There is no issue using AdminInterface

File: src/Admin/PageAdmin.php (hunk 2 above)
Comment (eerison) [reply src/Admin/PageAdmin.php]: hmmm I see, but do you have any idea where the route is defined for https://github.com/sonata-project/SonataPageBundle/blob/4.x/src/Controller/PageAdminController.php#L257 👀

File: src/Controller/PageAdminController.php (hunk 1 above)
Comment (eerison) [reply src/Controller/PageAdminController.php]: Ok it's not a problem to use the AdminInterface, it gonna work, but don't you think the AdminInterface is quite huge 👀

File: src/Admin/PageAdmin.php (hunk 2 above)
Comment (jordisala1991) [reply src/Admin/PageAdmin.php]: Here: https://github.com/sonata-project/SonataPageBundle/blob/4.x/src/Admin/PageAdmin.php#L65

Comment (jordisala1991) [reply src/Admin/PageAdmin.php]: The first parameter gets converted camelCase and added "Action" at the end, and that's how it knows the name.

File: src/Controller/PageAdminController.php [hunk 3]
```diff
@@ -38,6 +38,18 @@
...
+
+    public function __construct(
+        BlockServiceManagerInterface $blockServiceManager,
+        TemplateManagerInterface $templateManager
+    ) {
+        $this->blockServiceManager = $blockServiceManager;
+        $this->templateManager = $templateManager;
+
+    }
+
```
Comment (jordisala1991) [inline src/Controller/PageAdminController.php]: IMO, we should not change how we load things inside CRUDController, we use the container, with the function below `getSubscribedServices`

File: src/Controller/PageAdminController.php [hunk 4]
```diff
@@ -256,22 +268,18 @@ public function composeAction(Request $request): Response
      */
     public function composeContainerShowAction(Request $request): Response
     {
-        $blockAdmin = $this->container->get('sonata.page.admin.block');
-        \assert($blockAdmin instanceof BlockAdmin);
-
-        if (false === $blockAdmin->isGranted('LIST')) {
+        //TODO Can I use this directly from symfony controller?
+        if (false === $this->isGranted('LIST')) {
```
Comment (jordisala1991) [inline src/Controller/PageAdminController.php]: nope, this should remain as using blockAdmin, it generates the permission different. And IMO we should change for $blockAdmin->checkAccess('list') (but I don't see any relation with your bug report tbh, so it should be done in a separate PR).

File: src/Controller/PageAdminController.php [hunk 5]
```diff
@@ -256,22 +268,18 @@ public function composeAction(Request $request): Response
...
             throw new AccessDeniedException();
         }
 
         $id = $request->get($this->admin->getIdParameter());
-        $block = $blockAdmin->getObject($id);
+        $block = $this->admin->getObject($id);
```
Comment (jordisala1991) [inline src/Controller/PageAdminController.php]: This admin is a pageAdmin, not a block, it will return a page, this change is not correct.

File: src/Controller/PageAdminController.php (hunk 1 above)
Comment (jordisala1991) [reply src/Controller/PageAdminController.php]: Yes it is huge, but splitting it is a gigantic work, let's keep things simple for now 🙏

**Comment by jordisala1991** [comment]
> > What is not working? maybe if you add a failing test first it would be easier to see. (I tried the compose page quite a lot and fixed all the errors there)
> 
> the problem posted here: [#1495 (comment)](https://github.com/sonata-project/SonataPageBundle/issues/1495#issuecomment-1223876198)
> 
> and yes, I'm trying to add a failing test first, to find out what exactly is broken.

I would say you should revert all the `src` changes, and focus only on `tests` folder, otherwise it would be quite hard to identify the bug.

File: src/Admin/PageAdmin.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Admin/PageAdmin.php]: I think it's done with the name instead.

when you call `add->('compose_container_show')` the RouteCollection/RouteGenerator are automatically looking for a `composeContainerShow`
https://docs.sonata-project.org/projects/SonataAdminBundle/en/4.x/cookbook/recipe_custom_action/

File: src/Controller/PageAdminController.php (hunk 4 above)
Comment (eerison) [reply src/Controller/PageAdminController.php]: yeah it's not related with my issue, I'm just trying to make my test simple!

File: src/Admin/PageAdmin.php (hunk 2 above)
Comment (eerison) [reply src/Admin/PageAdmin.php]: > I think it's done with the name instead.
> 
> when you call `add->('compose_container_show')` the RouteCollection/RouteGenerator are automatically looking for a `composeContainerShow` https://docs.sonata-project.org/projects/SonataAdminBundle/en/4.x/cookbook/recipe_custom_action/

hmmmm I guess I got it :D, Thanks :)

File: src/Controller/PageAdminController.php (hunk 3 above)
Comment (eerison) [reply src/Controller/PageAdminController.php]: hmmm then don't we inject anything into the controllers' construct 👀

Comment (VincentLanglet) [reply src/Controller/PageAdminController.php]: We don't inject on the CRUDController because people are extending them, and we let them the constructor free.

Here the class is final so we could inject, but I agree that since it's already has a SubscribedServices we can keep using it for consistency.

File: src/Controller/PageAdminController.php (hunk 4 above)
Comment (VincentLanglet) [reply src/Controller/PageAdminController.php]: If you provide failing test when you changed src, we can't know if the test if failing because of the bug or because of your changes. So src code should not be changed.

**Comment by eerison** [comment]
it was solved in other PR :)
